src/main/java/org/jabref/model/entry/BibEntry.java

```diff
@@ -804,7 +805,8 @@ private void invalidateFieldCache(String fieldName) {
             return Optional.empty();
         }
 
-        return this.setField(FieldName.FILE, newValue);
+        //Violates our architecture, but currently only possible way because the LinkedFile is setup using JavaFX properties
+        return DefaultTaskExecutor.runInJavaFXThread(() -> this.setField(FieldName.FILE, newValue));
```

Comment:
that's a problem of our model then. I would vote to move the setup of the linked file to the model instead of violating the architecture in this way.

Reply:
There shouldn't be any problem with setting a JavaFX property on a non-FX thread. The properties don't check on which thread they are invoked (e.g. I think the basic example by oracle about the binding stuff actually uses a command line application). The problem occurs if you bind a JavaFX control to these properties, since then you get a cascade like
set javafx property > set some property in the control via binding > update view
and the last step is expected to be on the fx thread.

So far for the general case. Now here, where do we use bindings on the file field?

Reply:
In fact, updateItem on the LIstviewCellFactory is called which in the end leads to the thread error.
```

1:46:47.408 [Spin-2] ERROR org.jabref.FallbackExceptionHandler - Uncaught exception occurred in Thread[Spin-2,6,spin]
java.lang.IllegalStateException: Not on FX application thread; currentThread = Spin-2
	at com.sun.javafx.tk.Toolkit.checkFxUserThread(Toolkit.java:236) ~[jfxrt.jar:?]
	at com.sun.javafx.tk.quantum.QuantumToolkit.checkFxUserThread(QuantumToolkit.java:423) ~[jfxrt.jar:?]
	at javafx.scene.Parent$2.onProposedChange(Parent.java:367) ~[jfxrt.jar:?]
	at com.sun.javafx.collections.VetoableListDecorator.setAll(VetoableListDecorator.java:113) ~[jfxrt.jar:?]
	at com.sun.javafx.collections.VetoableListDecorator.setAll(VetoableListDecorator.java:108) ~[jfxrt.jar:?]
	at com.sun.javafx.scene.control.skin.LabeledSkinBase.updateChildren(LabeledSkinBase.java:575) ~[jfxrt.jar:?]
	at com.sun.javafx.scene.control.skin.LabeledSkinBase.handleControlPropertyChanged(LabeledSkinBase.java:181) ~[jfxrt.jar:?]
	at com.sun.javafx.scene.control.skin.ListCellSkin.handleControlPropertyChanged(ListCellSkin.java:49) ~[jfxrt.jar:?]
	at com.sun.javafx.scene.control.skin.BehaviorSkinBase.lambda$registerChangeListener$61(BehaviorSkinBase.java:197) ~[jfxrt.jar:?]
	at com.sun.javafx.scene.control.MultiplePropertyChangeListenerHandler$1.changed(MultiplePropertyChangeListenerHandler.java:55) ~[jfxrt.jar:?]
	at javafx.beans.value.WeakChangeListener.changed(WeakChangeListener.java:89) ~[jfxrt.jar:?]
	at com.sun.javafx.binding.ExpressionHelper$SingleChange.fireValueChangedEvent(ExpressionHelper.java:182) ~[jfxrt.jar:?]
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81) ~[jfxrt.jar:?]
	at javafx.beans.property.ObjectPropertyBase.fireValueChangedEvent(ObjectPropertyBase.java:105) ~[jfxrt.jar:?]
	at javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:112) ~[jfxrt.jar:?]
	at javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:146) ~[jfxrt.jar:?]
	at javafx.css.StyleableObjectProperty.set(StyleableObjectProperty.java:82) ~[jfxrt.jar:?]
	at javafx.beans.property.ObjectProperty.setValue(ObjectProperty.java:69) ~[jfxrt.jar:?]
	at javafx.scene.control.Labeled.setGraphic(Labeled.java:423) ~[jfxrt.jar:?]
	at org.jabref.gui.util.ViewModelListCellFactory$1.updateItem(ViewModelListCellFactory.java:122) ~[bin/:?]
	at javafx.scene.control.ListCell.updateItem(ListCell.java:480) ~[jfxrt.jar:?]
	at javafx.scene.control.ListCell.lambda$new$160(ListCell.java:167) ~[jfxrt.jar:?]
	at javafx.collections.WeakListChangeListener.onChanged(WeakListChangeListener.java:88) ~[jfxrt.jar:?]
	at com.sun.javafx.collections.ListListenerHelper$Generic.fireValueChangedEvent(ListListenerHelper.java:329) ~[jfxrt.jar:?]
	at com.sun.javafx.collections.ListListenerHelper.fireValueChangedEvent(ListListenerHelper.java:73) ~[jfxrt.jar:?]
	at javafx.collections.ObservableListBase.fireChange(ObservableListBase.java:233) ~[jfxrt.jar:?]
	at javafx.collections.ListChangeBuilder.commit(ListChangeBuilder.java:482) ~[jfxrt.jar:?]
	at javafx.collections.ListChangeBuilder.endChange(ListChangeBuilder.java:541) ~[jfxrt.jar:?]
	at javafx.collections.ObservableListBase.endChange(ObservableListBase.java:205) ~[jfxrt.jar:?]
	at com.sun.javafx.collections.ObservableListWrapper.remove(ObservableListWrapper.java:167) ~[jfxrt.jar:?]
	at com.sun.javafx.binding.BidirectionalContentBinding$ListContentBinding.onChanged(BidirectionalContentBinding.java:135) ~[jfxrt.jar:?]
	at com.sun.javafx.binding.ListExpressionHelper$Generic.notifyListeners(ListExpressionHelper.java:593) ~[jfxrt.jar:?]
	at com.sun.javafx.binding.ListExpressionHelper$Generic.fireValueChangedEvent(ListExpressionHelper.java:571) ~[jfxrt.jar:?]
	at com.sun.javafx.binding.ListExpressionHelper.fireValueChangedEvent(ListExpressionHelper.java:109) ~[jfxrt.jar:?]
	at javafx.beans.property.ListPropertyBase.fireValueChangedEvent(ListPropertyBase.java:200) ~[jfxrt.jar:?]
	at javafx.beans.property.ListPropertyBase.lambda$new$39(ListPropertyBase.java:56) ~[jfxrt.jar:?]
	at com.sun.javafx.collections.ListListenerHelper$SingleChange.fireValueChangedEvent(ListListenerHelper.java:164) ~[jfxrt.jar:?]
	at com.sun.javafx.collections.ListListenerHelper.fireValueChangedEvent(ListListenerHelper.java:73) ~[jfxrt.jar:?]
	at javafx.collections.ObservableListBase.fireChange(ObservableListBase.java:233) ~[jfxrt.jar:?]
	at javafx.collections.ListChangeBuilder.commit(ListChangeBuilder.java:482) ~[jfxrt.jar:?]
	at javafx.collections.ListChangeBuilder.endChange(ListChangeBuilder.java:541) ~[jfxrt.jar:?]
	at javafx.collections.ObservableListBase.endChange(ObservableListBase.java:205) ~[jfxrt.jar:?]
	at javafx.collections.ModifiableObservableListBase.setAll(ModifiableObservableListBase.java:90) ~[jfxrt.jar:?]
	at javafx.beans.binding.ListExpression.setAll(ListExpression.java:370) ~[jfxrt.jar:?]
	at org.jabref.gui.util.BindingsHelper.lambda$5(BindingsHelper.java:121) ~[bin/:?]
	at org.jabref.gui.util.BindingsHelper$BidirectionalListBinding.changed(BindingsHelper.java:192) ~[bin/:?]
	at com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(ExpressionHelper.java:361) ~[jfxrt.jar:?]
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81) ~[jfxrt.jar:?]
	at javafx.beans.property.StringPropertyBase.fireValueChangedEvent(StringPropertyBase.java:103) ~[jfxrt.jar:?]
	at javafx.beans.property.StringPropertyBase.markInvalid(StringPropertyBase.java:110) ~[jfxrt.jar:?]
	at javafx.beans.property.StringPropertyBase.set(StringPropertyBase.java:144) ~[jfxrt.jar:?]
	at javafx.beans.property.StringPropertyBase.set(StringPropertyBase.java:49) ~[jfxrt.jar:?]
	at javafx.beans.property.StringProperty.setValue(StringProperty.java:65) ~[jfxrt.jar:?]
	at javafx.beans.property.StringProperty.setValue(StringProperty.java:57) ~[jfxrt.jar:?]
	at org.jabref.gui.util.BindingsHelper$BidirectionalBinding.updateLocked(BindingsHelper.java:164) ~[bin/:?]
	at org.jabref.gui.util.BindingsHelper$BidirectionalBinding.changedB(BindingsHelper.java:157) ~[bin/:?]
	at com.sun.javafx.binding.ExpressionHelper$SingleChange.fireValueChangedEvent(ExpressionHelper.java:182) [jfxrt.jar:?]
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81) [jfxrt.jar:?]
	at javafx.beans.binding.ObjectBinding.invalidate(ObjectBinding.java:172) [jfxrt.jar:?]
	at com.sun.javafx.binding.BindingHelperObserver.invalidated(BindingHelperObserver.java:51) [jfxrt.jar:?]
	at com.sun.javafx.collections.MapListenerHelper$Generic.fireValueChangedEvent(MapListenerHelper.java:320) [jfxrt.jar:?]
	at com.sun.javafx.collections.MapListenerHelper.fireValueChangedEvent(MapListenerHelper.java:72) [jfxrt.jar:?]
	at com.sun.javafx.collections.ObservableMapWrapper.callObservers(ObservableMapWrapper.java:115) [jfxrt.jar:?]
	at com.sun.javafx.collections.ObservableMapWrapper.put(ObservableMapWrapper.java:169) [jfxrt.jar:?]
	at org.jabref.model.entry.BibEntry.setField(BibEntry.java:422) [bin/:?]
	at org.jabref.model.entry.BibEntry.setField(BibEntry.java:448) [bin/:?]
	at org.jabref.model.entry.BibEntry.setFiles(BibEntry.java:807) [bin/:?]
```

Reply:
I think the best solution is actually to call the whole cleanup process in the JavaFX thread.  If this does not work, you can add the Platform.runLater also in `org.jabref.gui.util.BindingsHelper`

Reply:
It works in that case that I do not receive any errors, but the file isn't renamed. I think the problem is that somehow it immediately returns instead of waiting